turbine/src/xdp.rs

```diff
@@ -0,0 +1,183 @@
+// re-export since this is needed at validator startup
+pub use agave_xdp::set_cpu_affinity;
+#[cfg(target_os = "linux")]
+use {
+    agave_xdp::{
+        device::{NetworkDevice, QueueId},
+        load_xdp_program,
+        tx_loop::tx_loop,
+    },
+    crossbeam_channel::TryRecvError,
+    std::{sync::Arc, thread::Builder, time::Duration},
+};
+use {
+    crossbeam_channel::{Sender, TrySendError},
+    solana_ledger::shred,
+    std::{error::Error, net::SocketAddr, thread},
+};
+
+#[derive(Clone, Debug)]
+pub struct XdpConfig {
+    pub interface: Option<String>,
+    pub cpus: Vec<usize>,
+    pub zero_copy: bool,
+    pub channel_cap: usize,
+}
+
+impl XdpConfig {
+    // A nice round number
+    const DEFAULT_CHANNEL_CAP: usize = 1_000_000;
+}
+
+impl Default for XdpConfig {
+    fn default() -> Self {
+        Self {
+            interface: None,
+            cpus: vec![],
+            zero_copy: false,
+            channel_cap: Self::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+impl XdpConfig {
+    pub fn new(interface: Option<impl Into<String>>, cpus: Vec<usize>, zero_copy: bool) -> Self {
+        Self {
+            interface: interface.map(|s| s.into()),
+            cpus,
+            zero_copy,
+            channel_cap: XdpConfig::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        shred_index: u32,
+        addr: Vec<SocketAddr>,
+        payload: shred::Payload,
+    ) -> Result<(), TrySendError<(Vec<SocketAddr>, shred::Payload)>> {
+        self.senders[shred_index as usize % self.senders.len()].try_send((addr.clone(), payload))
+    }
+}
+
+pub(crate) struct XdpRetransmitter {
+    threads: Vec<thread::JoinHandle<()>>,
+}
+
+impl XdpRetransmitter {
+    #[cfg(not(target_os = "linux"))]
+    pub(crate) fn new(
+        _config: XdpConfig,
+        _src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        Err("XDP is only supported on Linux".into())
+    }
+
+    #[cfg(target_os = "linux")]
+    pub(crate) fn new(
+        config: XdpConfig,
+        src_port: u16,
+    ) -> Result<(Self, XdpSender), Box<dyn Error>> {
+        use caps::{
+            CapSet,
```

Comment:
Can you elaborate how the capability will work here? Would we have to start the whole of agave under cap_net_admin?

Reply:
The process requires cap_net_admin and cap_bpf as _permitted_ capabilities. When XDP is enabled, the process _temporarily_ assumes those capabilities, does what it has to do, then drops them immediately after. So other than briefly at startup, the validator runs with the same capabilities as today.

Reply:
But can it then reaquire these caps or can we permanently drop them?

Reply:
The XDP threads drop them permanently. The other threads can in theory acquire them, but we don't even have the code for that, so it's not going to magically happen.

Because of the way we setup stages and threads etc, it's non trivial to shuffle things around so that eg XDP initialization is done outside of RetransmitStage. It's possible, but it would be a huge change, and as we established this PR is already too big, so I will gladly let you do that work after we land this PR!

Reply:
Ok fair point, lets set this into backlog.